bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/IndexPersistenceMgr.java

```diff
@@ -220,7 +220,19 @@ CachedFileInfo getFileInfo(final Long ledger, final byte masterKey[]) throws IOE
                 } else {
                     fi = readFileInfoCache.get(ledger, loader);
                 }
-            } while (!fi.tryRetain());
+                if (!fi.tryRetain()) {
+                    // defensively ensure that dead fileinfo objects don't exist in the
+                    // cache. They shouldn't if refcounting is correct, but if someone
+                    // does a double release, the fileinfo will be cleaned up, while
+                    // remaining in the cache, which could cause a tight loop in this method.
+                    if (writeFileInfoCache.asMap().remove(ledger, fi)
+                        || readFileInfoCache.asMap().remove(ledger, fi)) {
+                        LOG.error("Dead fileinfo({}) forced out of cache. Must have been double-released somewhere.",
```

Comment:
Is there a way to automatically recover from this situation ? if not I would prefer to crash the bookie and not to run with a broken cache, what do you think @ivankelly  ?

Reply:
automatic recover is removal from the caches. The fileinfo object is dead, so nothing can be done with it in any case.